pkg/sidecar/k8s_instance.go

```diff
@@ -0,0 +1,386 @@
+	if online && ((cfg.IPv6 != nil && !link.IPv6.IP.Equal(cfg.IPv6.IP)) ||
+		(cfg.IPv4 != nil && !link.IPv4.IP.Equal(cfg.IPv4.IP))) {
+		// Disconnect and reconnect to change the IP addresses.
+		logging.S().Debug("disconnect and reconnect to change the IP addr", "cfg.IPv4", cfg.IPv4, "link.IPv4", link.IPv4.String(), "container", n.container.ID)
+		//
+		// NOTE: We probably don't need to do this on local docker.
+		// However, we probably do with swarm.
+		online = false
+		if err := n.cninet.DelNetworkList(ctx, link.netconf, link.rt); err != nil {
+			return fmt.Errorf("when 5: %w", err)
+		}
+		delete(n.activeLinks, cfg.Network)
+	}
+
+	// Are we _connected_ to the network.
+	if !online {
+		// No, we're not.
+		// Connect.
+		if cfg.IPv6 != nil {
+			return errors.New("ipv6 not supported")
+		}
+
+		var (
+			netconf *libcni.NetworkConfigList
+			err     error
+		)
+		if cfg.IPv4 == nil {
+			netconf, err = newNetworkConfigList("net", n.subnet)
+		} else {
+			netconf, err = newNetworkConfigList("ip", cfg.IPv4.String())
+		}
+		if err != nil {
+			return fmt.Errorf("failed to generate new network config list: %w", err)
+		}
+
+		cniArgs := [][2]string{}                   // empty
+		capabilityArgs := map[string]interface{}{} // empty
+		ifName := "eth1"
+
+		rt := &libcni.RuntimeConf{
+			ContainerID:    n.container.ID,
+			NetNS:          n.netnsPath,
+			IfName:         ifName,
+			Args:           cniArgs,
+			CapabilityArgs: capabilityArgs,
+		}
+
+		_, err = n.cninet.AddNetworkList(ctx, netconf, rt)
+		if err != nil {
+			return fmt.Errorf("failed to add network through cni plugin: %w", err)
+		}
+
+		netlinkByName, err := n.nl.LinkByName(ifName)
+		if err != nil {
+			return fmt.Errorf("failed to get link by name: %w", err)
+		}
+
+		// Register an active link.
+		handle, err := NewNetlinkLink(n.nl, netlinkByName)
+		if err != nil {
+			return fmt.Errorf("failed to register new netlink: %w", err)
+		}
+		v4addrs, err := handle.ListV4()
+		if err != nil {
+			return fmt.Errorf("failed to list v4 addrs: %w", err)
+		}
+		if len(v4addrs) != 1 {
+			return fmt.Errorf("expected 1 v4addrs, but received %d", len(v4addrs))
+		}
+
+		link = &k8sLink{
+			NetlinkLink: handle,
+			IPv4:        v4addrs[0],
+			IPv6:        nil,
+			rt:          rt,
+			netconf:     netconf,
+		}
+
+		logging.S().Debugw("successfully adding an active link", "ipv4", link.IPv4, "container", n.container.ID)
+
+		n.activeLinks[cfg.Network] = link
+	}
+
+	// We don't yet support applying per-subnet rules.
+	if len(cfg.Rules) != 0 {
+		return fmt.Errorf("TODO: per-subnet bandwidth rules not supported")
+	}
+
+	if err := link.Shape(cfg.Default); err != nil {
+		return fmt.Errorf("failed to shape link: %w", err)
+	}
+	return nil
+}
+
+func (n *K8sNetwork) ListActive() []string {
+	networks := make([]string, 0, len(n.activeLinks))
+	for name := range n.activeLinks {
+		networks = append(networks, name)
+	}
+	return networks
+}
+
+func (n *K8sNetwork) ListAvailable() []string {
```

Comment:
We can drop this.

Reply:
Not sure what you mean here - this is required by the interface.

Reply:
We can drop it from the interface as well. I think I pushed a patch to do that but it looks like it got clobbered.

ListActive is useful. I don't have any immediate uses for ListAvailable.